Add snapshot fork feature

PatelUtkarsh opened the pull request:
Fixes: #41 
Any UI improvements are welcome. 

Remaining:
- [x] Post meta Meta fork/copy 
- [x] Should append (fork) in a post title or not confirm and update.
- [x] Unit-test

PatelUtkarsh committed:
Add snapshot fork feature

PatelUtkarsh committed:
Copy meta of  fork post

PatelUtkarsh committed:
Forked post states add for post list table

PatelUtkarsh committed:
do not override post object

PatelUtkarsh committed:
Fork meta-box add parent snapshot info

PatelUtkarsh committed:
Add unit test for snapshot-fork :white_check_mark:

PatelUtkarsh committed:
Rename snapshot 'Fork/Copy' to 'Fork'

PatelUtkarsh committed:
Add more testcase for snapshot fork :large_blue_circle:

PatelUtkarsh committed:
Remove post-status restriction for fork

PatelUtkarsh commented:
@westonruter / @valendesigns This PR is ready for review.
This will conflict with #84 and I will be happy to resolve conflicts.

valendesigns commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,35 @@
+
+	component.forkClick = function() {
+		var $forkButton = $( '#snapshot-fork' ),
+			$forkSnipper = $( '.snapshot-fork-spinner' ),
```
We could change this to `$spinner` or at the very least fix the typo and change it to `$forkSpinner`.

valendesigns commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,35 @@
+		var $forkButton = $( '#snapshot-fork' ),
+			$forkSnipper = $( '.snapshot-fork-spinner' ),
+			$forkList = $( '#snapshot-fork-list' ),
+			forkItemTemplate = wp.template( 'snapshot-fork-item' );
```
Add return space.

valendesigns commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,35 @@
+			request = wp.ajax.post( 'snapshot_fork', {
+				ID: $forkButton.data( 'post-id' ),
+				nonce: $forkButton.data( 'nonce' )
+			} );
```
Add return space.

valendesigns commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,35 @@
+			} );
+			request.always( function() {
+				$forkSnipper.removeClass( 'is-active' );
+			} );
```
Add return space.

valendesigns commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,35 @@
+				var item;
+				item = $( $.trim( forkItemTemplate( data ) ) );
+				$forkList.append( item );
+			} );
```
Add return space.

valendesigns commented on php/class-customize-snapshot-manager.php:
```diff
@@ -747,6 +748,23 @@ public function enqueue_frontend_scripts() {
+	 */
+	public function enqueue_admin_scripts( $hook ) {
+		global $post;
+		$handle = 'customize-snapshots-admin';
```
This could be moved into the conditional, and add return space after the global.

valendesigns commented on php/class-post-type.php:
```diff
@@ -341,11 +353,15 @@ public function render_data_metabox( $post ) {
 			$frontend_view_url = get_permalink( $post->ID );
 			echo sprintf(
-				'<a href="%s" class="button button-secondary">%s</a>',
+				'<a href="%s" class="button button-secondary">%s</a> ',
```
Is the extra whitespace in this line change a typo?

PatelUtkarsh replied:
It is a change.

valendesigns replied:
If it's just to add visual space to the buttons then it should probably be done with CSS on `#snapshot-fork` instead.

PatelUtkarsh replied:
Just being consistent with https://github.com/xwp/wp-customize-snapshots/blob/0c20ece186044ad94a0d381c394c15d4030b83c4/php/class-post-type.php#L349

Should I remove space between both button and add CSS?

valendesigns commented on php/class-post-type.php:
```diff
@@ -322,6 +332,8 @@ public function render_data_metabox( $post ) {
 		echo sprintf( '%1$s %2$s %3$s', esc_html__( 'Modified:', 'customize-snapshots' ), esc_html( get_the_modified_date( '' ) ), esc_html( get_the_modified_time( '' ) ) ) . '<br>';
 		echo '</p>';
 
+		$fork_markup = '<a href="#" id="snapshot-fork" class="button button-secondary" data-post-id="' . $post->ID . '" data-nonce="' . wp_create_nonce( 'snapshot-fork' ) . '">' . esc_html__( 'Fork', 'customize-snapshots' ) . '</a>';
```
This should be a button not a link and also using `sprintf` would clean things up. Something like:

```
$fork_markup = sprintf(
    '<button id="snapshot-fork" class="button button-secondary" data-post-id="%s" data-nonce="%s">%s</button><span class="spinner snapshot-fork-spinner"></span>',
    $post->ID,
    wp_create_nonce( 'snapshot-fork' ),
    esc_html__( 'Fork', 'customize-snapshots' )
);
```

valendesigns commented on php/class-post-type.php:
```diff
@@ -416,6 +432,30 @@ public function render_data_metabox( $post ) {
+		</ul><?php
+		if ( $post->post_parent ) {
+			$parent = get_post( $post->post_parent );
+			echo '<h2>' . __( 'Parent:', 'customize-snapshots' ) . '&nbsp;<a href="' . esc_url( get_edit_post_link( $parent, 'raw' ) ) . '">' . get_the_title( $parent ) . '</a></h2>';
```
Could probably remove the `&nbsp;` and just add a space.

PatelUtkarsh committed:
Code review changes

PatelUtkarsh committed:
Revert space

PatelUtkarsh committed:
Merge develop in 'feature/snapshot-fork'

PatelUtkarsh committed:
Fix issue with fork metabox post_status; improve code coverage

PatelUtkarsh committed:
Fix phpcs

westonruter commented on js/customize-snapshots-admin.js:
```diff
@@ -0,0 +1,39 @@
+			component.forkClick();
+		} );
+	};
+	component.init();
```
This nice component isn't exposed anywhere for the world to see. I suggest `return component;` here, and to not do self-booting script. So the top of this script can be like:

``` js
var CustomizeSnapshotsAdmin = (function( $, wp ) {
```

And then when enqueueing the script, also do:

``` php
wp_add_inline_script( 'customize-snapshots-admin', 'CustomizeSnapshotsAdmin.init()' );
```

This will allow the JS to be more readily extended and tested.

westonruter commented on php/class-post-type.php:
```diff
@@ -822,4 +867,71 @@ public function admin_show_merge_error() {
+	 * Handles snapshot fork ajax request.
+	 */
+	public function handle_snapshot_fork() {
+		if ( ! check_ajax_referer( 'snapshot-fork', 'nonce', false ) ) {
```
This should probably do a capability check for good measure.

PatelUtkarsh committed:
Code review changes

PatelUtkarsh commented:
I've updated code in f5b86b1 
Also replaced action `admin_print_footer_scripts-post.php` with `admin_footer-post.php` as ealier action was not available in WP 4.5.4.

westonruter commented on php/class-post-type.php:
```diff
@@ -822,4 +867,77 @@ public function admin_show_merge_error() {
+		                           ||
+		                           ! intval( $_POST['ID'] )
+		                           ||
+		                           ! current_user_can( $post_type_object->cap->edit_post, $_POST['ID'] )
```
`intval( $_POST['ID'] )` should be passed into the `current_user_can()` check instead of a bare `$_POST['ID']`. As it stands right now, this is passing an unsanitized post value into `current_user_can()`. For example, if `$_POST['ID']` is `'234abc'` then this will run through `intval()` as `234` and pass that check, but then it will get passed in as that unsanitized string to the `current_user_can()`.

You should also make sure that there actually exists a post for the given post and that the post type matches. Looks like you're doing that down below as well, but I think the post's existence and type should be verified before it gets run through the cap check.

You may consider breaking up this condition into multiple `if` statements and then moving `$post_id = intval( $_POST['ID'] );` up so that you can re-use `$post_id`.

PatelUtkarsh committed:
Snapshot fork: Change request authentication order and fix unsanitize check

PatelUtkarsh committed:
Merge 'develop' in 'feature/snapshot-fork' (Resolve conflicts and refactor customize-snapshots-admin.js)

PatelUtkarsh committed:
Merge 'refactor/intergrate-changeset' into 'feature/snapshot-fork'

PatelUtkarsh committed:
Merge branch 'refactor/intergrate-changeset' into feature/snapshot-fork

PatelUtkarsh committed:
Merge branch 'refactor/intergrate-changeset' into feature/snapshot-fork

PatelUtkarsh committed:
Merge branch 'refactor/intergrate-changeset' into feature/snapshot-fork

PatelUtkarsh committed:
Fix fork issue back compatibility

PatelUtkarsh committed:
Merge branch 'refactor/intergrate-changeset' into feature/snapshot-fork

PatelUtkarsh committed:
Merge branch 'develop' into feature/snapshot-fork

PatelUtkarsh committed:
Merge branch 'develop' into feature/snapshot-fork

 Conflicts:
	php/class-post-type.php

PatelUtkarsh committed:
:hammer: Fix phpcs

PatelUtkarsh committed:
Merge branch 'develop' into feature/snapshot-fork

valendesigns commented:
@PatelUtkarsh Please fix merge conflicts so I can review. Thanks!

PatelUtkarsh committed:
Merge branch 'develop' into feature/snapshot-fork

 Conflicts:
	tests/php/test-class-post-type.php

PatelUtkarsh committed:
Merge branch 'develop' into feature/snapshot-fork

PatelUtkarsh commented:
Todo:
- [x] Test
- [x] Style @sayedtaqui

mohdsayed committed:
Code refactoring and improvements

mohdsayed committed:
Update phpunit test for refactoring

mohdsayed committed:
Fix phpunit test for handle_snapshot_fork

mohdsayed committed:
Code refactoring of static properites

mohdsayed committed:
Add styling to forked snapshots link

PatelUtkarsh committed:
Fix issue with remove setting

PatelUtkarsh commented:
Looks good, Ready for review. @westonruter

PatelUtkarsh committed:
Merge branch 'develop' into feature/snapshot-fork

PatelUtkarsh committed:
Fix phpcs

westonruter committed:
Focus on newly-added forked snapshot in metabox after clicking fork button

* Add fork button to Changeset Forks metabox.
* Increase number of forked changesets listed, with showing message if more than 100 exist.

westonruter committed:
Add missing wp_slash() for insertion of post and postmeta

Improve jsdoc

westonruter committed:
Fix covers tag in test_force_at_least_one_revision

westonruter committed:
Update dev-lib and disable coveralls since composer package missing

westonruter commented:
It appears that the coveralls Composer package is broken. I can see:

> Skipped installation of bin bin/coveralls for package satooshi/php-coveralls: file not found in package

I'll disable it.
